### PR TITLE
[Utils] Adds `store()` and `restore()` methods to EMAModel

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR adds the `store()` and `restore()` methods to our `EMAModel` class. The reason is beautifully explained by @patil-suraj [here](https://github.com/huggingface/diffusers/pull/2173#discussion_r1091895090). 

An end-to-end example of how to use these methods in practice is [here](https://gist.github.com/sayakpaul/075ee224a5332779710f6b649bfa304d#file-train_text_to_image-py-L853-L902). 

It also un-blocks https://github.com/huggingface/diffusers/pull/2157

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Ok for me! Let's just be a bit careful when adding this functionality to the training scripts to not blow up GPU memory

File: src/diffusers/training_utils.py [hunk 1]
```diff
@@ -251,6 +251,32 @@ def state_dict(self) -> dict:
...
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        r"""
+        Args:
+        Save the current parameters for restoring later.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored.
+        """
+        parameters = list(parameters)
+        self.collected_params = [param.clone() for param in parameters]
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:262]:
```suggestion
        self.collected_params = [param.detach().cpu().clone() for param in parameters]
```

Think we should force to store parameters on CPU here

Comment (patrickvonplaten) [reply src/diffusers/training_utils.py:262]: This way it doesn't get in the way of blowing up GPU memory

**Comment by sayakpaul** [comment]
For the code quality tests, I am running `make style` and `make quality`. It picks up a script unrelated to this PR:

* `src/diffusers/training_utils.py`

My formatting virtual environment is up-to-date as well:

* black: 23.1.0
* isort: 5.12.0
* ruff: 0.0.244

**Comment by sayakpaul** [comment]
Still the same issue: https://github.com/huggingface/diffusers/pull/2302#issuecomment-1428060576

**Comment by patrickvonplaten** [comment]
Fixed the `make quality` issue. @sayakpaul, your hf-doc-builder library might be outdated. 

We also use https://github.com/huggingface/doc-builder to improve the style of our code (think it makes sure comments are not too long. 

You can install it with:

```
pip install hf-doc-builder
```

File: src/diffusers/training_utils.py [hunk 2]
```diff
@@ -251,6 +251,32 @@ def state_dict(self) -> dict:
...
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        r"""
+        Args:
+        Save the current parameters for restoring later.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored.
+        """
+        parameters = list(parameters)
+        self.collected_params = [param.detach().cpu().clone() for param in parameters]
```
Comment (williamberman) [inline src/diffusers/training_utils.py:262]: can we just have the caller do the `detach.cpu.clone` and save collected_params in a variable themselves? This sort of adhoc storage just between two methods is a bit of an OO antipattern imo

File: src/diffusers/training_utils.py [hunk 3]
```diff
@@ -251,6 +251,32 @@ def state_dict(self) -> dict:
...
+        Args:
+        Restore the parameters stored with the `store` method. Useful to validate the model with EMA parameters without:
+        affecting the original optimization process. Store the parameters before the `copy_to()` method. After
+        validation (or model saving), use this to restore the former parameters.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                updated with the stored parameters. If `None`, the parameters with which this
+                `ExponentialMovingAverage` was initialized will be used.
+        """
+        if self.collected_params is None:
+            raise RuntimeError("This ExponentialMovingAverage has no `store()`ed weights " "to `restore()`")
```
Comment (williamberman) [inline src/diffusers/training_utils.py:275]: If we do https://github.com/huggingface/diffusers/pull/2302/files#r1107495110, then this method wouldn't need to reference self and it wouldn't have this error condition. This would mean maybe it would look better as a function, and then store and restore could be free-standing functions not on a class.

WDYT?

**Review by williamberman — APPROVED** [review]
I think these shouldn't be methods on the class but ok with merging if other people feel strongly it should be :)

File: src/diffusers/training_utils.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/training_utils.py:275]: Since the `EMAModel` class majorly going to be used in the examples, which are meant to be easier to work with, I like this design approach better. So, with that in mind, for https://github.com/huggingface/diffusers/pull/2302#discussion_r1107495110, I think it's probably better if we do it ourselves. 

From what I have seen is that the community is fairly eager to customize things as per their needs, but with that, we should aim to provide example scripts that are easy as starting points. 

WDYT?

File: src/diffusers/training_utils.py (hunk 2 above)
Comment (williamberman) [reply src/diffusers/training_utils.py:262]: Another thing is I think the initial `list()` is also unnecessary. If something is an Iterable, I think we should be able to directly iterate over it?

Comment (sayakpaul) [reply src/diffusers/training_utils.py:262]: Yes, that is true. The detach code was after, hence.

File: src/diffusers/training_utils.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/training_utils.py:275]: I think we can still make the methods easy to work with without the adhoc variable storage. I.e.

```py
# training utils
def store(params):
    return [x.cpu().detach() for x in params]

def restore(params, stored_params):
    for c_param, param in zip(stored_params, parameters):
            param.data.copy_(c_param.data)


...

# training script
stored = store(params)
# do something
restore(params, stored)
```

is still friendly and hackable without the adhoc variable storage

Comment (sayakpaul) [reply src/diffusers/training_utils.py:275]: Then it is coming down to a preference now :D 

I like the class approach better than you showed above.

File: src/diffusers/training_utils.py [hunk 4]
```diff
@@ -251,6 +251,32 @@ def state_dict(self) -> dict:
...
+        parameters = list(parameters)
```
Comment (williamberman) [inline src/diffusers/training_utils.py:276]:
```suggestion
```

File: src/diffusers/training_utils.py [hunk 5]
```diff
@@ -251,6 +251,32 @@ def state_dict(self) -> dict:
...
         }
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        r"""
+        Args:
+        Save the current parameters for restoring later.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored.
+        """
+        parameters = list(parameters)
```
Comment (williamberman) [inline src/diffusers/training_utils.py:261]:
```suggestion
```

File: src/diffusers/training_utils.py [hunk 6]
```diff
@@ -251,6 +251,30 @@ def state_dict(self) -> dict:
             "collected_params": self.collected_params,
         }
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:254]:
```suggestion
    def store_non_ema(self, parameters: Iterable[torch.nn.Parameter]) -> None:
```

File: src/diffusers/training_utils.py [hunk 7]
```diff
@@ -251,6 +251,30 @@ def state_dict(self) -> dict:
...
+        r"""
+        Args:
+        Save the current parameters for restoring later.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored.
+        """
+        self.collected_params = [param.detach().cpu().clone() for param in parameters]
+
+    def restore(self, parameters: Iterable[torch.nn.Parameter]) -> None:
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:263]:
```suggestion
    def restore_non_ema(self, parameters: Iterable[torch.nn.Parameter]) -> None:
```

File: src/diffusers/training_utils.py [hunk 8]
```diff
@@ -251,6 +251,30 @@ def state_dict(self) -> dict:
...
         }
 
+    def store(self, parameters: Iterable[torch.nn.Parameter]) -> None:
+        r"""
+        Args:
+        Save the current parameters for restoring later.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                temporarily stored.
+        """
+        self.collected_params = [param.detach().cpu().clone() for param in parameters]
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:261]: (nit) Why is it called `collected_params`? Might it be better to call it `temp_stored_params`?

**Review by patrickvonplaten — APPROVED** [review]
I understand @williamberman points here and to be honest the `EMAModel` class is also not super intuitive to me in general. To me it's actually more of a `EMAModelTrainingUtils` class. 

But from a user point of view that uses the examples training scripts I also think it's more convenient to be able to attach temporarily stored parameters to the `EMAModel` instead of having some [`stored` list flying around](https://github.com/huggingface/diffusers/pull/2302/files#r1107960983) 

Maybe some renaming could help here a bit to make it super clear that the `EMAModel.store(...)` function actually does not store and restore ema parameters but the non_ema parameters?

File: src/diffusers/training_utils.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/diffusers/training_utils.py:261]: But it's a nit so happy to merge without

**Comment by patrickvonplaten** [comment]
Also good to go for me, but would be happy about potentially renaming the functions a bit to make the difference between the actual EMA and the non-EMA weights clearer

File: src/diffusers/training_utils.py (hunk 8 above)
Comment (patil-suraj) [inline src/diffusers/training_utils.py:261]: Not really sure why we need to put this on CPU and clone. `store`, `restore` shouldn't really take extra memory as we already have both params on GPU. Keeping this on the CPU and then loading it back on GPU will introduce some slowdown.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR. I agree with Patrick regarding naming. 
I just left a comment about copying the params CPU in store. Not sure if it's really needed, will need to dig into it a bit. The rest of it looks great!

**Comment by sayakpaul** [comment]
I am in agreement regarding the name changes. Applying them now. I will let @patil-suraj play around with the CPU offloading (`.cpu()`) before then.

File: src/diffusers/training_utils.py (hunk 8 above)
Comment (patil-suraj) [reply src/diffusers/training_utils.py]: okay play around with it a bit, and think this is okay, sorry about the false alarm.  eval is only once a epoch/few epochs so this shouldn't really be a big issue. Although we should leave a note in the training script that for large models and max perf the user should feel free to modify this.

cc @sayakpaul @patrickvonplaten

File: src/diffusers/training_utils.py [hunk 9]
```diff
@@ -298,11 +322,11 @@ def load_state_dict(self, state_dict: dict) -> None:
...
-            if len(self.collected_params) != len(self.shadow_params):
-                raise ValueError("collected_params and shadow_params must have the same length")
+        self.temp_stored_params = state_dict.get("temp_stored_params", None)
+        if self.temp_stored_params is not None:
+            if not isinstance(self.temp_stored_params, list):
+                raise ValueError("temp_stored_params must be a list")
+            if not all(isinstance(p, torch.Tensor) for p in self.temp_stored_params):
+                raise ValueError("temp_stored_params must all be Tensors")
+            if len(self.temp_stored_params) != len(self.shadow_params):
+                raise ValueError("temp_stored_params and shadow_params must have the same length")
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:332]: Actually, there's no need to save and load these temp params, they are only really necessary during validation and we don't need to keep them after that.

File: src/diffusers/training_utils.py [hunk 10]
```diff
@@ -248,9 +248,33 @@ def state_dict(self) -> dict:
...
+        affecting the original optimization process. Store the parameters before the `copy_to()` method. After
+        validation (or model saving), use this to restore the former parameters.
+            parameters: Iterable of `torch.nn.Parameter`; the parameters to be
+                updated with the stored parameters. If `None`, the parameters with which this
+                `ExponentialMovingAverage` was initialized will be used.
+        """
+        if self.temp_stored_params is None:
+            raise RuntimeError("This ExponentialMovingAverage has no `store()`ed weights " "to `restore()`")
+        for c_param, param in zip(self.temp_stored_params, parameters):
+            param.data.copy_(c_param.data)
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:276]: Once we are done restoring, we could just del the temp params, not really necessary to keep them around.

**Comment by patil-suraj** [comment]
We could do this in another PR, but it would be nice to add some tests for ema model. @sayakpaul feel free to work on it if you want.

File: src/diffusers/training_utils.py (hunk 9 above)
Comment (sayakpaul) [reply src/diffusers/training_utils.py:332]: Do you suggest deleting these after the validation?

**Comment by sayakpaul** [comment]
> We could do this in another PR, but it would be nice to add some tests for ema model. @sayakpaul feel free to work on it if you want.

I can but in a separate PR.

File: src/diffusers/training_utils.py (hunk 9 above)
Comment (patil-suraj) [reply src/diffusers/training_utils.py:332]: Yes, in the `restore` method, once we the params are restored we can delete the temp params and no need to save/load these in the state dict.

Comment (sayakpaul) [reply src/diffusers/training_utils.py:332]: We can do `del self.temp_stored_params` but then again in the subsequent iterations, that would lead to errors since `restore()` is called in intervals. Am I missing out on anything?

Comment (patil-suraj) [reply src/diffusers/training_utils.py:332]: `store` and `restore` should always be called in sequence. It doesn't make sense to call `restore` without `store`. So if it leads to an error that's better. And by delete I mean we could simply set it to `None` after restoring is done.

Comment (sayakpaul) [reply src/diffusers/training_utils.py:332]: Yeah makes sense now! Will proceed now.

File: src/diffusers/training_utils.py [hunk 11]
```diff
@@ -149,7 +149,7 @@ def save_pretrained(self, path):
         model = self.model_cls.from_config(self.model_config)
         state_dict = self.state_dict()
         state_dict.pop("shadow_params", None)
-        state_dict.pop("collected_params", None)
+        state_dict.pop("temp_stored_params", None)
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:152]: No need to save this in `state_dict`.

File: src/diffusers/training_utils.py [hunk 12]
```diff
@@ -298,11 +324,11 @@ def load_state_dict(self, state_dict: dict) -> None:
...
-            if len(self.collected_params) != len(self.shadow_params):
-                raise ValueError("collected_params and shadow_params must have the same length")
+        self.temp_stored_params = state_dict.get("temp_stored_params", None)
+        if self.temp_stored_params is not None:
+            if not isinstance(self.temp_stored_params, list):
+                raise ValueError("temp_stored_params must be a list")
+            if not all(isinstance(p, torch.Tensor) for p in self.temp_stored_params):
+                raise ValueError("temp_stored_params must all be Tensors")
+            if len(self.temp_stored_params) != len(self.shadow_params):
+                raise ValueError("temp_stored_params and shadow_params must have the same length")
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:334]: This can be removed, as saving temp params is not necessary. 
```suggestion
```

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for updating this quickly. Just left two quick comments, we can remove the temp params from saving/loading state_dict.

**Comment by sayakpaul** [comment]
I think the error related to running `fix-copies` is very likely from an unrelated issue.

**Comment by patil-suraj** [comment]
The `EMAModel` is in the main init, and hence it's also in the dummy objects (https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/dummy_pt_objects.py#L618). Whenever we update any object in the main init we need to run `make fix-copies`.

**Comment by sayakpaul** [comment]
> The `EMAModel` is in the main init, and hence it's also in the dummy objects (https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/dummy_pt_objects.py#L618). Whenever we update any object in the main init we need to run `make fix-copies`.

Right. Sorry for missing it out. Should be good now.

**Comment by patil-suraj** [comment]
Thanks! The failing test is unrelated, merging this now:)
